osc-server/src/test/java/org/osc/core/broker/service/tasks/conformance/openstack/sfc/UpdateServiceFunctionChainTaskTest.java

```diff
@@ -16,17 +16,170 @@
+import org.osc.core.broker.model.sdn.NetworkElementImpl;
+import org.osc.core.broker.util.TransactionalBroadcastUtil;
+import org.osc.core.broker.util.db.DBConnectionManager;
+import org.osc.core.test.util.TestTransactionControl;
+import org.osc.core.test.util.mockito.matchers.NetworkElementIdMatcher;
+import org.osc.sdk.controller.api.SdnRedirectionApi;
+import org.osc.sdk.controller.element.NetworkElement;
 
-@Ignore
+@RunWith(MockitoJUnitRunner.class)
 public class UpdateServiceFunctionChainTaskTest {
 
+    @Rule
+    public ExpectedException exception = ExpectedException.none();
+
+    @Mock
+    private EntityManager em;
+
+    @Mock
+    private EntityTransaction tx;
+
+    @Mock(answer = Answers.CALLS_REAL_METHODS)
+    private TestTransactionControl txControl;
+
+    @Mock
+    private DBConnectionManager dbMgr;
+
+    @Mock
+    private TransactionalBroadcastUtil txBroadcastUtil;
+
+    @Mock
+    private ApiFactoryService apiFactoryServiceMock;
+
+    @Mock
+    private SdnRedirectionApi sdnApi;
+
+    @Captor
+    private ArgumentCaptor<List<NetworkElement>> neListCaptor;
+
+    @InjectMocks
+    private UpdateServiceFunctionChainTask task;
+
+    private SecurityGroup sg;
+    private ServiceFunctionChain sfc;
+    private List<NetworkElement> networkElementList;
+
+    @Before
+    public void testInitialize() throws Exception {
+        MockitoAnnotations.initMocks(this);
+
+        Mockito.when(this.em.getTransaction()).thenReturn(this.tx);
+
+        this.txControl.setEntityManager(this.em);
+
+        Mockito.when(this.dbMgr.getTransactionalEntityManager()).thenReturn(this.em);
+        Mockito.when(this.dbMgr.getTransactionControl()).thenReturn(this.txControl);
+
+        VirtualizationConnector mockVc = mock(VirtualizationConnector.class);
+        this.sg = new SecurityGroup(mockVc, "PROJECT_ID", "PROJECT_NAME");
+        this.sg.setId(2L);
+        this.sg.setName("A_SECURITY_GROUP");
+
+        Mockito.when(this.em.find(SecurityGroup.class, this.sg.getId())).thenReturn(this.sg);
+
+        this.sfc = new ServiceFunctionChain();
+        this.sfc.setId(3L);
+        this.sfc.setName("A_SFC_NAME");
+
+        Mockito.when(this.em.find(ServiceFunctionChain.class, this.sfc.getId())).thenReturn(this.sfc);
+
+        Mockito.when(this.apiFactoryServiceMock.createNetworkRedirectionApi(mockVc)).thenReturn(this.sdnApi);
+
+    }
+
     @Test
-    public void test() {
-        fail("Not yet implemented");
+    public void testExecuteTransaction_WithSecurityGroupSFCIdSetToNull_ExpectUpdate() throws Exception {
+        // Arrange
+        UpdateServiceFunctionChainTask updateTask = this.task.create(this.sfc, this.sg, this.networkElementList);
+        Mockito.when(this.sdnApi
+                .updateNetworkElement(argThat(new NetworkElementIdMatcher(this.sg.getNetworkElementId())), any()))
+                .thenReturn(new NetworkElementImpl("UPDATED_SFC_ID"));
+
+        assertNull(this.sg.getNetworkElementId());
```

Comment:
no need for this assertion, this is basically testing the test code which is redundant.

Reply:
making sure that the task execution actually updates a null field...if something does change in the testInitialize we are making sure the update still happens correctly

Reply:
IMO the test code should be able to trust the test code, making assumptions in the test code should be ok and it keeps each method simple and targeted. As long as we keep dependencies light and most things local trusting it should be ok, cross checking becomes a necessity if the test code has many non local dependencies. If we cannot trust the test code itself does it mean we always need to double check the initialize? It could get complicated. Also, ideally all tests can follow the recipe arrange, act, assert. If we add asserts in the arrange part it can get confusing.

Reply:
Another way to make the test more explicit while removing the assert in the arrange phase would be to simply do this.sg.setNe...(null). So no matter what the initialize does this test is setting it to null, it may also make the test more explicit.